include/ibamr/BrinkmanPenalizationAdvDiff.h

```diff
@@ -0,0 +1,306 @@
+     * conditions.
+     *
+     * \note Inhomogeneous Neumann BCs are treated uniquely within this class and require
+     * additional user callback inputs.
+     */
+    void registerInhomogeneousNeumannBC(SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q_var,
+                                        SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > ls_solid_var,
+                                        BrinkmanInhomogeneousNeumannBCsFcnPtr callback,
+                                        void* ctx,
+                                        double num_interface_cells = 2.0,
+                                        double eta_penalty_coef = 1.0e-4);
+
+    /*!
+     * \brief Function to compute the cell-centered coefficient to the damping linear operator
+     * and RHS of the advection-diffusion equation for a specified transported quantity Q_var
+     *
+     * \note It is assumed that the physical damping coefficient \f$\lambda\f$ is zero.
+     *
+     * The functional form of the Brinkman damping coefficient is
+     * \f$ C = \sum_{Dirichlet} \chi_i/\eta \f$ where \f$\chi_i = 1-H_i\f$ and the sum is taken over all level sets with
+     * Dirichlet BCs. Note that Neumann BCs do not contribute anything to this term.
+     *
+     */
+    void computeDampingCoefficient(int C_idx, SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q_var);
+
+    /*!
+     * \brief Function to compute the side-centered coefficient to the diffusion linear operator
+     * and RHS of the advection-diffusion equation for a specified transported quantity Q_var with diffusion
+     * coefficient kappa.
+     *
+     * \note This function is able to handle both constant and variable kappa.
+     *
+     * The functional form of the Brinkman diffusion coefficient is
+     * \f$ D = \kappa + \sum_{Neumann} (-\chi_i + \eta \chi_i)\f$ where \f$\chi_i = 1-H_i\f$ and the sum is taken over
+     * all level sets with Neumann BCs. Note that Dirichlet BCs do not contribute anything to this term.
+     */
+    void computeDiffusionCoefficient(int D_idx,
+                                     SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q_var,
+                                     int kappa_idx,
+                                     double kappa);
+
+    /*!
+     * \brief Function to compute the Brinkman forcing contribution to the RHS of the advection-diffusion solver
+     * for a specified transported quantity Q_var.
+     *
+     * For Dirichlet BCs, \f$ F_i = \chi_i/\eta Q_{bc}\f$ where \f$\chi_i = 1-H_i\f$.
+     * For homogeneous Neumann BCs, \f$ F_i = 0\f$.
+     * For inhomogeneous Neumann BCs, \f$ F_i = \nabla \dot (\chi_i B) - \chi_i \nabla \dot B_i \f$, with a user defined
+     * \f$B_i\f$.
+     * The overall functional form of the Brinkman body force is \f$F = \sum_{i} F_i\f$.
+     */
+    void computeForcing(int F_idx, SAMRAI::tbox::Pointer<SAMRAI::pdat::CellVariable<NDIM, double> > Q_var);
```

Comment:
For instance, in the analogy to IB for FSI, this forcing could be passed to the advection-diffusion solver though the `CartGridFunction` interface without having to tell the `AdvDiffHierarchyIntegrator` class about the Brinkmann class.

Reply:
When I made this comment I did not realize that this involved changing the linear operator, and so this suggestion is not relevant.